File: content/LLD/Reference-Code/18. Proxy Pattern/bad_example.py
import time


class HighResImage:
    def __init__(self, file_name: str):
        self.__file_name = file_name
        self.__image_data = None

        self._load_from_disk()

    def _load_from_disk(self):
        time.sleep(1)
        self.__image_data = f"[[Loaded image data of {self.__file_name}]]"
        print(f"Loaded {self.__file_name} from disk")

    def display(self):
        print(f"{self.__file_name} has image data of {self.__image_data}\n")


class PhotoGallery:
    def __init__(self):
        self.__images: list[HighResImage] = []

    def add_image(self, file_name: str):
        high_res_image = HighResImage(file_name)
        self.__images.append(high_res_image)

    def display_gallery(self):
        for image in self.__images:
            image.display()


photo_gallery = PhotoGallery()
photo_gallery.add_image("image1.png")
photo_gallery.add_image("image2.png")
photo_gallery.add_image("image3.png")
photo_gallery.add_image("image4.png")

# photo_gallery.display_gallery()

File: content/LLD/Reference-Code/18. Proxy Pattern/BadExample.java
import java.util.ArrayList;
import java.util.List;

public class BadExample {
    static class HighResImage {
        private final String fileName;
        private String imageData;

        public HighResImage(String fileName) {
            this.fileName = fileName;
            loadFromDisk();
        }

        private void loadFromDisk() {
            try {
                Thread.sleep(50); // Simulating expensive I/O
            } catch (InterruptedException ignored) {}
            this.imageData = "[[Loaded image data of " + this.fileName + "]]";
            System.out.println("Loaded " + this.fileName + " from disk");
        }

        public void display() {
            System.out.println(this.fileName + " has image data of " + this.imageData + "\n");
        }
    }

    static class PhotoGallery {
        private final List<HighResImage> images = new ArrayList<>();

        public void addImage(String fileName) {
            HighResImage highResImage = new HighResImage(fileName);
            this.images.add(highResImage);
        }

        public void displayGallery() {
            for (HighResImage image : this.images) {
                image.display();
            }
        }
    }

    public static void main(String[] args) {
        PhotoGallery photoGallery = new PhotoGallery();
        photoGallery.addImage("image1.png");
        photoGallery.addImage("image2.png");
        photoGallery.addImage("image3.png");
        photoGallery.addImage("image4.png");
    }
}
